### PR TITLE
Add new complex template application test from issue #24

### PR DESCRIPTION
**Pull request by jhdavis8**
See issue #24. This test checks that we can declare a complicated templated struct within a target region. It does not check for any particular results, so it will not intentionally give any errors as long as it runs to completition. I am not very familiar with template metaprogramming so additional eyes to make sure this makes sense is greatly appreciated.

Also included are some whitespace fixes for other tests in this folder.

### CODE REVIEW

File: tests/4.5/application_kernels/complex_template.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
+//===--- complex_template.c ------------------------------------------------===//
+//
+// OpenMP API Version 4.5 Nov 2015
+//
+// This is a stripped-down application code which sets up two template
+// structs which use standard library functions. A templated struct
+// containing the other struct is declared on the device. This test does
+// not check any results as it is only a test of declaring a templated
+// struct on the device.
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/complex_template.cpp]: Add credit to original application

File: tests/4.5/application_kernels/complex_template.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+int main(int argc, char *argv[]) {
+  OMPVV_TEST_OFFLOADING;
+
+  OMPVV_TEST_SHARED_ENVIRONMENT;
+
+  int errors = 0;
+
+#pragma omp target
+  {
+    S<0> s;
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/complex_template.cpp]: Test more combinations for the template parameter (unsigned long, something not zero)

Comment (tmh97) [reply tests/4.5/application_kernels/complex_template.cpp]: We may need to make the template slightly simpler to be able to test this properly

**Comment by jhdavis8** [comment]
Right now the first test, for the first constructor, is passing, but the second check is not. I will be investigating this further.

File: tests/4.5/application_kernels/complex_template.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  int errors = 0;
+  int version_set[2] = {-1};
+
+#pragma omp target map(from: version_set[0:2])
+  {
+    S<0> s;
+    version_set[0] = s.v.version_called;
+    R<0> r;
+    r.v = *(new V<0>((unsigned long) 1));
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/complex_template.cpp]: Change the Dim in the Template to something other than 0.

File: tests/4.5/application_kernels/complex_template.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
+//===--- complex_template.cpp -----------------------------------------------===//
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/complex_template.cpp]: Rename file to include Alpaka first.

**Comment by jhdavis8** [comment]
This updated version is passing xlc++ and g++ but does not compile on clang with a long error message that ends with a request to file a bug, so I believe that is the appropriate next course of action. Otherwise, this test is ready for review.

File: tests/4.5/application_kernels/alpaka_complex_template.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  }
+};
+
+template<int Dim> struct S {
+  V<Dim> v;
+};
+
+template<int Dim> struct R {
+  V<Dim> v;
+};
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/alpaka_complex_template.cpp]: Remove this struct, not used

Comment (jhdavis8) [reply tests/4.5/application_kernels/alpaka_complex_template.cpp]: Add full error message to PR

**Comment by jhdavis8** [comment]
Clang error message:
clang-3.8: /home/compteam/slave1/build_folder_ibm_rpm_c_92/src/tools/clang/lib/CodeGen/CodeGenFunction.h:3542: void clang::CodeGen::CodeGenFunction::setAddrOfLocalVar(const clang::VarDecl *, clang::CodeGen::Addres
s): Assertion `!LocalDeclMap.count(VD) && "Decl already exists in LocalDeclMap!"' failed.
#0 0x00000000119b14d8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x119b14d8)
#1 0x00000000119b1a30 PrintStackTraceSignalHandler(void*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x119b1a30)
#2 0x00000000119b1d10 SignalHandler(int) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x119b1d10)
#3 0x00002000000504d8  0x4d8 __GI_abort
#4 0x00002000000504d8 
#5 0x00002000000504d8 __assert_fail_base (+0x4d8)
#6 0x0000200000472094 __GI___assert_fail (/usr/lib64/libc.so.6+0x42094)
#7 0x00002000004656d4 clang::CodeGen::CodeGenFunction::setAddrOfLocalVar(clang::VarDecl const*, clang::CodeGen::Address) (/usr/lib64/libc.so.6+0x356d4)
#8 0x00002000004657c4 clang::CodeGen::CodeGenFunction::GenerateOpenMPCapturedStmtFunction(clang::CapturedStmt const&, bool, unsigned int, unsigned int, bool) (/usr/lib64/libc.so.6+0x357c4)
#9 0x0000000011bdd454 clang::CodeGen::CGOpenMPRuntime::outlineTargetDirective(clang::OMPExecutableDirective const&, llvm::StringRef, clang::CodeGen::RegionCodeGenTy const&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20
180531/bin/clang-3.8+0x11bdd454)
#10 0x0000000011bdadec clang::CodeGen::CGOpenMPRuntime::emitTargetOutlinedFunctionHelper(clang::OMPExecutableDirective const&, llvm::StringRef, llvm::Function*&, llvm::Constant*&, bool, clang::CodeGen::RegionCodeG
enTy const&, unsigned int) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11bdadec)
#11 0x0000000011e2d0b4 clang::CodeGen::CGOpenMPRuntime::emitTargetOutlinedFunction(clang::OMPExecutableDirective const&, llvm::StringRef, llvm::Function*&, llvm::Constant*&, bool, clang::CodeGen::RegionCodeGenTy c
onst&, unsigned int) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11e2d0b4)
#12 0x0000000011e2d47c emitCommonOMPTargetDirective(clang::CodeGen::CodeGenFunction&, clang::OMPExecutableDirective const&, clang::OpenMPDirectiveKind, clang::CodeGen::RegionCodeGenTy const&, llvm::ArrayRef<llvm::
Value*>, clang::CodeGen::OMPMapArrays&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11e2d47c)
#13 0x0000000011e2ce50 clang::CodeGen::CodeGenFunction::EmitOMPTargetDirective(clang::OMPTargetDirective const&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11e2ce50)
#14 0x0000000011bf6e14 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11bf6e14)
#15 0x0000000011bf6840 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x1
1bf6840)
#16 0x0000000011bc8580 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11bc85
80)
#17 0x0000000011bd1bec clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11bd1bec)
#18 0x0000000011c1c364 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11c1c364)
#19 0x0000000011c3563c clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11c3563c)
#20 0x0000000011c2f7d4 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11c2f7d4)
#21 0x0000000011c33440 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11c33440)
#22 0x0000000011c38270 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11c38270)
#23 0x000000001242d44c clang::ParseAST(clang::Sema&, bool, bool) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x1242d44c)
#24 0x000000001242b524 clang::ASTFrontendAction::ExecuteAction() (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x1242b524)
#25 0x000000001296f8e0 clang::CodeGenAction::ExecuteAction() (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x1296f8e0)
#26 0x0000000011fe24dc clang::FrontendAction::Execute() (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11fe24dc)
#27 0x000000001242a05c clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x1242a05c)
#28 0x0000000011fe1c90 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11fe1c90)
#29 0x0000000011f9e2e4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x11f9e2e4)
#30 0x00000000120841c8 main (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x120841c8)
#31 0x0000000010465920 generic_start_main.isra.0 (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x10465920)
#32 0x00000000104630b4 __libc_start_main (/autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8+0x104630b4)
0  clang-3.8 0x00000000119b14d8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 120
1  clang-3.8 0x00000000119b1a30
2  clang-3.8 0x00000000119b1d10
3            0x00002000000504d8 __kernel_sigtramp_rt64 + 0
4  libc.so.6 0x0000200000472094 abort + 692
5  libc.so.6 0x00002000004656d4                                                                                                                                                                             [10/1832]
6  libc.so.6 0x00002000004657c4 __assert_fail + 100
7  clang-3.8 0x0000000011bdd454
8  clang-3.8 0x0000000011bdadec clang::CodeGen::CodeGenFunction::GenerateOpenMPCapturedStmtFunction(clang::CapturedStmt const&, bool, unsigned int, unsigned int, bool) + 748
9  clang-3.8 0x0000000011e2d0b4 clang::CodeGen::CGOpenMPRuntime::outlineTargetDirective(clang::OMPExecutableDirective const&, llvm::StringRef, clang::CodeGen::RegionCodeGenTy const&) + 532
10 clang-3.8 0x0000000011e2d47c clang::CodeGen::CGOpenMPRuntime::emitTargetOutlinedFunctionHelper(clang::OMPExecutableDirective const&, llvm::StringRef, llvm::Function*&, llvm::Constant*&, bool, clang::CodeGen::Re
gionCodeGenTy const&, unsigned int) + 668
11 clang-3.8 0x0000000011e2ce50 clang::CodeGen::CGOpenMPRuntime::emitTargetOutlinedFunction(clang::OMPExecutableDirective const&, llvm::StringRef, llvm::Function*&, llvm::Constant*&, bool, clang::CodeGen::RegionCo
deGenTy const&, unsigned int) + 80
12 clang-3.8 0x0000000011bf6e14
13 clang-3.8 0x0000000011bf6840 clang::CodeGen::CodeGenFunction::EmitOMPTargetDirective(clang::OMPTargetDirective const&) + 528
14 clang-3.8 0x0000000011bc8580 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*) + 896
15 clang-3.8 0x0000000011bd1bec clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) + 172
16 clang-3.8 0x0000000011c1c364 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) + 1572
17 clang-3.8 0x0000000011c3563c clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) + 540
18 clang-3.8 0x0000000011c2f7d4 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) + 276
19 clang-3.8 0x0000000011c33440 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) + 1744
20 clang-3.8 0x0000000011c38270 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) + 272
21 clang-3.8 0x000000001242d44c
22 clang-3.8 0x000000001242b524
23 clang-3.8 0x000000001296f8e0 clang::ParseAST(clang::Sema&, bool, bool) + 544
24 clang-3.8 0x0000000011fe24dc clang::ASTFrontendAction::ExecuteAction() + 220
25 clang-3.8 0x000000001242a05c clang::CodeGenAction::ExecuteAction() + 284
26 clang-3.8 0x0000000011fe1c90 clang::FrontendAction::Execute() + 144
27 clang-3.8 0x0000000011f9e2e4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 2036
28 clang-3.8 0x00000000120841c8 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 4840
29 clang-3.8 0x0000000010465920 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 1952
30 clang-3.8 0x00000000104630b4 main + 12564
31 libc.so.6 0x0000200000455200
32 libc.so.6 0x00002000004553f4 __libc_start_main + 196
Stack dump:
0.      Program arguments: /autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/clang-3.8 -cc1 -triple powerpc64le-unknown-linux-gnu -emit-llvm-bc -emit-llvm-uselists -disable-free -main-file-name alpaka_complex_temp
late.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -fuse-init-array -target-cpu ppc64le -mfloat-abi hard -target-abi elfv2 -dwarf-c
olumn-info -debugger-tuning=gdb -resource-dir /autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/../lib/clang/3.8.0 -I ./ompvv -D __NO_MATH_INLINES -U __SSE2_MATH__ -U __SSE_MATH__ -D VERBOSE_MODE=1 -I/sw/summit/cu
da/9.2.148/include -c-isystem /sw/summit/llvm/1.0-20180531/omprtl -cxx-isystem /sw/summit/llvm/1.0-20180531/omprtl -internal-isystem /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5 -internal-
isystem /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/ppc64le-redhat-linux -internal-isystem /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/backward -internal-isystem
 /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5 -internal-isystem /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/ppc64le-redhat-linux -internal-isystem /usr/lib/gcc/pp
c64le-redhat-linux/4.8.5/../../../../include/c++/4.8.5/backward -internal-isystem /usr/local/include -internal-isystem /autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/../lib/clang/3.8.0/include -internal-externc
-isystem /include -internal-externc-isystem /usr/include -internal-isystem /usr/local/include -internal-isystem /autofs/nccs-svm1_sw/summit/llvm/1.0-20180531/bin/../lib/clang/3.8.0/include -internal-externc-isyste
m /include -internal-externc-isystem /usr/include -O3 -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /ccs/home/jhdavis/sollve-project/sollve_vv -ferror-limit 19 -fmessage-length 0 -fopenmp -fno-signed-char 
-fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -vectorize-loops -vectorize-slp -o /tmp/alpaka_complex_template-0a07cd.bc -x c++ /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/
tests/4.5/application_kernels/alpaka_complex_template.cpp -fopenmp-targets=nvptx64-nvidia-cuda 
1.      <eof> parser at end of file
2.      /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/4.5/application_kernels/alpaka_complex_template.cpp:39:5: LLVM IR generation of declaration 'main'
3.      /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/4.5/application_kernels/alpaka_complex_template.cpp:39:5: Generating code for declaration 'main'
clang-3.8: error: unable to execute command: Aborted (core dumped)
clang-3.8: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.8.0 (ibmgithub:/CORAL-LLVM-Compilers/clang.git 7d4a8c43e2c34864ffff52cd62d3a604b048d6cf) (ibmgithub:/CORAL-LLVM-Compilers/llvm.git aa08e5a3c3670cd86fb4bee034a7626bb26ad57e)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/llvm/1.0-20180531/bin
clang-3.8: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang-3.8: note: diagnostic msg: Error generating preprocessed source(s).

**Comment by jhdavis8** [comment]
Not sure if a bug should be reported for this test since one has already been reported and fixed for Clang 3.8.0. The test passes fine on gcc and xlc, as well as Clang 9.0.0.

**Comment by jhdavis8** [comment]
Passing on all summit compilers (Clang 9.0.0)

**Pull request merged**